athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/storage/StorageMetadata.java

```diff
@@ -0,0 +1,314 @@
+            Optional<FieldValue> optionalFieldValue = FieldValue.from(path);
+            if (optionalFieldValue.isEmpty()) {
+                continue;
+            }
+            FieldValue fieldValue = optionalFieldValue.get();
+            Optional<EqualsExpression> optionalExpression = expressions.stream()
+                    .filter(expr -> expr.columnName().equalsIgnoreCase(fieldValue.getField()))
+                    .findFirst();
+            if (optionalExpression.isPresent()) {
+                LOGGER.info("Evaluating field value {} against the expression {}", fieldValue, expressions);
+                EqualsExpression expression = optionalExpression.get();
+                if (!expression.apply(fieldValue.getValue())) {
+                    return false;
+                }
+            }
+            else {
+                LOGGER.info("No expression found for field {}", fieldValue.getField());
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Builds the table schema based on the provided field
+     *
+     * @param table      Glue table object
+     * @return An instance of {@link Schema}
+     */
+    public Schema buildTableSchema(Table table, BufferAllocator allocator) throws URISyntaxException
+    {
+        SchemaBuilder schemaBuilder = SchemaBuilder.newBuilder();
+        String locationUri = table.getStorageDescriptor().getLocation();
+        URI storageLocation = new URI(locationUri);
+        List<Field> fieldList = getFields(storageLocation.getAuthority(), storageLocation.getPath(), table.getParameters().get(CLASSIFICATION_GLUE_TABLE_PARAM), allocator);
+
+        LOGGER.debug("Schema Fields\n{}", fieldList);
+        for (Field field : fieldList) {
+            if (isFieldTypeNull(field)) {
+                field = Field.nullable(field.getName().toLowerCase(), Types.MinorType.VARCHAR.getType());
+            }
+            else {
+                field = new Field(field.getName().toLowerCase(), new FieldType(field.isNullable(), field.getType(), field.getDictionary(), field.getMetadata()), field.getChildren());
+            }
+            schemaBuilder.addField(getCompatibleField(field));
+        }
+        return schemaBuilder.build();
+    }
+
+    private Field getCompatibleField(Field field)
+    {
+        String fieldName = field.getName().toLowerCase();
+        Types.MinorType fieldType = Types.getMinorTypeForArrowType(field.getType());
+        switch (fieldType) {
+            case TIMESTAMPNANO:
+            case TIMESTAMPSEC:
+            case TIMESTAMPMILLI:
+            case TIMEMICRO:
+            case TIMESTAMPMICRO:
+            case TIMESTAMPMILLITZ:
+            case TIMESTAMPMICROTZ:
+            case TIMENANO:
+                    return new Field(fieldName,
+                            new FieldType(field.isNullable(), Types.MinorType.DATEMILLI.getType(), field.getDictionary(),
+                                    field.getMetadata()), field.getChildren());
```

Comment:
As I mentioned during the meeting, I'm fine with you guys wanting to do conversion, but there is a correctness issue here.

If we are getting 1000 nanoseconds, you can't just change the type to milliseconds without converting the value as well, right?

Similarly, for the timestamps with timezones, you will need to convert these values properly to UTC millis.